utils/staking-miner/src/main.rs

```diff
@@ -0,0 +1,523 @@
+// Copyright 2021 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+// ## Staking Miner
+//
+// things to look out for:
+// 1. weight (already taken care of).
+// 2. length (already taken care of).
+// 3. Important, but hard to do: memory usage of the chain.
+
+mod rpc_helpers;
+mod dry_run;
+mod emergency_solution;
+mod monitor;
+mod prelude;
+
+use jsonrpsee_ws_client::{WsClient, WsClientBuilder};
+use prelude::*;
+use remote_externalities::{Builder, Mode, OnlineConfig};
+use sp_core::crypto::Pair as _;
+use sp_runtime::traits::Block as BlockT;
+use std::path::{Path, PathBuf};
+use structopt::StructOpt;
+
+pub(crate) enum AnyRuntime {
+	Polkadot,
+	Kusama,
+	Westend,
+}
+
+pub(crate) static mut RUNTIME: AnyRuntime = AnyRuntime::Polkadot;
+
+macro_rules! construct_runtime_prelude {
+	($runtime:ident) => { paste::paste! {
+		#[allow(unused_import)]
+		pub(crate) mod [<$runtime _runtime_exports>] {
+			pub(crate) use crate::prelude::EPM;
+			pub(crate) use [<$runtime _runtime>]::*;
+			pub(crate) use crate::monitor::[<monitor_cmd_ $runtime>] as monitor_cmd;
+			pub(crate) use crate::dry_run::[<dry_run_cmd_ $runtime>] as dry_run_cmd;
+			pub(crate) use crate::emergency_solution::[<emergency_solution_cmd_ $runtime>] as emergency_solution_cmd;
+			pub(crate) use private::{[<create_uxt_ $runtime>] as create_uxt};
+
+			mod private {
+				use super::*;
+				pub(crate) fn [<create_uxt_ $runtime>](
+					raw_solution: EPM::RawSolution<EPM::CompactOf<Runtime>>,
+					witness: u32,
+					signer: crate::Signer,
+					nonce: crate::prelude::Index,
+					tip: crate::prelude::Balance,
+					era: sp_runtime::generic::Era,
+				) -> UncheckedExtrinsic {
+					use codec::Encode as _;
+					use sp_core::Pair as _;
+					use sp_runtime::traits::StaticLookup as _;
```

Comment:
I don't understand this name aliasing, but trusting it is correct :P

Reply:
This is a pattern when you just want to use a trait's function (and thereby need to `use` it), but don't want the trait name itself to be in scope.